readme
Jau


changelog.md
#Changelog

##Version 1.2.2 (aktuell)
* Veröffentlicht am 09. Mai 2015
* Optische Überarbeitung der Visitenkarte (-> Visitenkarte wird nun auch korrekt in der Stelzliste angezeigt)
* Schematische Auszeichung der Visitenkarte wurde verbessert

##Version 1.2.1
* Veröffentlicht am 02. Mai 2015
* Entfernung von Debug-Ausgaben in der Browserkonsole (console.log)
* Optimierung der Erfassung des aktuellen Pr0mium-Status für die "stelz"-Funktion
* Weitere Optimierungen beim Stelzen

##Version 1.2.0
* Veröffentlicht am 02. Mai 2015
* Die Visitenkarte wird nun schon angezeigt, sobald sich die Maus über dem Namen befindet (Hovereffekt)
* Es wurde ein Stelz-Button hinzugefügt (nutzbar für alle Pr0mium Benutzer)
* Solange die Daten aus dem Profil geladen werden, wird nun eine Ladeanimation angezeigt
* Bei einem Klick auf den Namen öffnet sich nicht mehr das Profil, sondern es werden man gelangt direkt zum Profil
* Es wurden im Skript zusätzliche Einstellungsmöglichkeiten hinzugefügt
* Die Visitenkartenanimation beim Öffnen wurde beschleunigt
* Die Daten des Profils werden nun in Echtzeit per AJAX geladen
* Bei gesperrten Benutzer wird kein Status mehr angezeigt, lediglich die Sperrdauer
* Optimierungen und Veränderungen im CSS-Code
* Optimierungen im Skript selbst

##Version 1.1.0
* Veröffentlicht am 28. April 2015
* Aktueller Rang/Punkt wird in der Visitenkarte nun dargestellt
* Es werden alle Badges angezeigt (vorher fehlten die Kommentar-, sowie die Jahresbadge)
* Alle Badges verweisen auf die richtige Seite
* Die Visitenkarte ist nun auch in der Inbox (Posteingang), sowie in der Stelzliste vorhanden
* Es wurden Skript interne Optimierungen vorgenommen
* Der Abstand zum nächsten Upload wurde von 95 Pixeln auf 105 Pixeln vergrößert (siehe readme.md)


##Version 1.0.0
* Veröffentlicht am 25. April 2015
* Release


gistfile1.js
// ==UserScript==

// @name        [Pr0] Visitenkarte
// @author  rebrain
// @namespace rebrain
// @copyright 2015
// @description Erzeugt eine Visitenkarte bei der Anzeige von Benutzernamen
// @version 1.0.0
// @grant none
// @include *pr0gramm.com*
// @icon  http://pr0gramm.com/media/pr0gramm-favicon.png
// @run-at  document-start
// @updateURL
// @downloadURL
// ==/UserScript==

var cacheName = 'profileCache';
var previousProfile = null;

$(function() {
    // Build cache
    createProfileCache();
    
    // CSS code
    var css = '.wide-button { width: 100%; padding-top: 5px; padding-bottom: 5px; margin-top: 5px; margin-bottom: 5px; } \
.user.profile { display: inline; } \
.user.profile > .name { cursor: pointer; } \
.user.profile > .extended { z-index: 9000; display: none; position: absolute; margin-top: 5px; min-width: 200px; padding: 5px; padding-top: 10px; background-color: rgba(0, 0, 0, 0.8); border: 1px solid rgb(42, 46, 49); border-radius: 2px; } \
.user.profile > .extended.loading { display: none; }  \
.user.profile > .extended > .data-row { display: inline; font-family: sans-serif; } \
.user.profile > .extended > .data-row:after { clear: both; display: block; content: " "; padding-bottom: 3px; } \
.user.profile > .extended > .data-row > .left { padding-right: 10px; color: #BBBBBB; float: left; } \
.user.profile > .extended > .data-row > .right { padding-left: 10px; float: right; } \
.user.profile > .extended > .data-row > .right.fliese { color: #F5F7F6; } \
.user.profile > .extended > .data-row > .right.rebrain { color: #EE4D2E; } \
.comments > .comment-box:last-child { margin-bottom: 95px; } \
';
    
    // Build item template hack
    var templateReplace = /(<a href="#user\/{item.user}" class="user um{item.mark}">{item.user}<\/a>)/;
    var templateExtension = '<div class="user profile um{item.mark}" data-showable="0" data-visible="0"><span class="name">{item.user}</span>\
<div class="extended" data-user="{item.user}"><span><a href="#user\/{item.user}" class="confirm-button wide-button">zum ausf&uuml;hrlichen Profil</a></span></div></div>';
    
    // Template hack for items
    var templateOld = p.View.Stream.Item.prototype.template;
    var templateNew = templateOld.replace(templateReplace, templateExtension);
    
    // Set new template and prevent endless bug
    p.View.Stream.Item = p.View.Stream.Item.extend({
		template: templateNew,
		show: function(rowIndex, itemData, defaultHeight, jumpToComment) {
			this.parent(rowIndex, itemData, defaultHeight, jumpToComment);
            previousProfile = null;
    }});
    
    // Build comment template hack
    var templateReplace = /<a href="#user\/{c.name}" class="user um{c.mark}">{c.name}<\/a>/;
    var templateExtension = '<div class="user profile um{c.mark}" data-showable="0" data-visible="0"><span class="name">{c.name}</span>\
<div class="extended" data-user="{c.name}"><span><a href="#user\/{c.name}" class="confirm-button wide-button">zum ausf&uuml;hrlichen Profil</a></div></div>';
    
    // Template hack for comments
    templateOld = p.View.Stream.Comments.prototype.template;
    templateNew = templateOld.replace(templateReplace, templateExtension);
    
    p.View.Stream.Comments.prototype.template = templateNew;
    
    // Now bind action to item and comment
    $(document).on('click', 'div.user.profile[data-showable][data-visible]', function() { handleUserProfile(this); });
    
    // CSS stuff
    $('head').append('<style type="text/css">' + css + '</style>');
    
});

function createProfileCache() {
    $('body').append('<div id="' + cacheName + '" hidden="hidden"></div>');
}

function isProfileCached(user) {
    return ($('body > div#' + cacheName + ' > div[data-user="' + user + '"]').length > 0);
}

function addProfileInCache(user, score, registration, bann, badges) {
    var content = '<div data-type="cache" data-user="' + user + '">';
    
    // Handle color
    if (score > 0) {
        var color = 'rebrain';
    } else {
        var color = 'fliese';
    }
    
    // Score
    content = content + createCacheLine('Benis', score, color);
    
    // Registration
    if (score < 0) {
        content = content + createCacheLine('geschrumpft seit', registration, color);
    } else {
        content = content + createCacheLine('gewachsen seit', registration, color);
    }
    
    // Bann
    if (bann != '') {
        content = content + createCacheLine('gesperrt f&uuml;r', bann, color);
    }
    
    // Badges
    if (badges != '') {
        content = content + createCacheLine('', badges, color);
    }
    
    // Profile link
    content = content + '</div>';
    
    // Provide in cache
    $('body > div#' + cacheName).append(content);
}

function createCacheLine(left, right, color) {
    return '<div class="data-row"><span class="left">' + left + '</span><span class="right ' + color + '">' + right + '</span></div>';
}

function loadProfileFromCache(user, target) {
    // Try to load from cache
    var cache = $('body > div#' + cacheName + ' > div[data-type="cache"][data-user="' + user + '"]');
    if (cache.length > 0) {
        target.children('div.extended').children('span').before(cache.html());
    }
    
    // If we could not load it from cache: we just show a link to full profile 
}

function loadProfileFromAPI(user) {
    // Build api url
    var profileAPI = '/api/profile/info?name=' + user;
    
    // AJAX request
    $.ajax({
        url: profileAPI,
        async: false,
        beforeSend: function() {
            $(document).css('cursor', 'progress');
        },
        complete: function() {
            $(document).css('cursor', undefined);
        },
        success: function(data) {
            // Handle data
            data = jQuery.parseJSON(data);
            var user = data.user;
            var badges = data.badges;
            
            // Score and registration
            var score = user.score;
            var since = (new Date (user.registered * 1000)).relativeTime(true, true);
            
            // Bann
            var banned = '';
            if (user.banned) {
                if (user.bannedUntil == null) {
                    banned = 'immer';
                } else {
                    banned = new Date(user.bannedUntil * 1000);
                    banned = banned.relativeTime(false, true);
                }
            }
            
            // Badges
            var badge = '';
            for (var i = 0; i < badges.length; i++) {
                badge = badge + ' <a class="badge" href="' + badges[i].link + '" title="' + badges[i].description + '"><img src="/media/badges/' + badges[i].image + '" class="badge" alt="" /></a>';
            }
            badge = badge.trim();
            
            // Provide to cache
            addProfileInCache(user.name, score, since, banned, badge);
        }
    });
}

function handleUserProfile(element) {
    // Get node
    element = $('body').find(element);
    
    // Hide item
    if (element.attr('data-visible') == 1) {
        // Close physically
        previousProfile = null;
        
        // Close visually
        element.children('div.extended').hide('slow');
        element.attr('data-visible', '0');
        
    // Show item
    } else {
        // Close previous profile
        if (previousProfile != null) {
            handleUserProfile(previousProfile);
        }
        
        // Load item
        if (element.attr('data-showable') != 1) {
            // Get username
            var user = element.children('div.extended').attr('data-user');
            
            // Load from api if required
            if (!isProfileCached(user)) {
                loadProfileFromAPI(user);
            }
            
            // Load from local cache
            loadProfileFromCache(user, element);
            element.attr('data-showable', '1');
        }
        
        // Mark current as visible
        previousProfile = element;
        
        // Now we are able to show
        element.children('div.extended').show('slow').css('display', 'inline');
        element.attr('data-visible', '1');
    }
}

QuickProfile.js
// ==UserScript==
// @name	[Pr0] Visitenkarte
// @author	rebrain
// @namespace	rebrain
// @copyright	2015
// @description	Erzeugt eine Visitenkarte bei der Anzeige von Benutzernamen
// @version	1.2.2
// @grant	none
// @include	*pr0gramm.com*
// @icon	http://pr0gramm.com/media/pr0gramm-favicon.png
// @updateURL	https://raw.githubusercontent.com/rebrainding/pr0/master/QuickProfile.js
// @downloadURL	https://raw.githubusercontent.com/rebrainding/pr0/master/QuickProfile.js
// ==/UserScript==

// HIER KANNST DU EINSTELLUNGEN VORNEHMEN UND DAS SKRIPT
// DEINEN WÜNSCHEN ANPASSEN, BEACHTE DIE ANMERKUNGEN!

// Geschwindigkeit mit der sich die Visitenkarte öffnet (erlaubte Werte: "fast", "normal", "slow")
var profileAnimationSpeed = 'fast';

// Dauer die das Profil mindestens angezeigt wird, sofern es geöffnet wurde (Wert in Millisekunden)
var profileMinShowDuration = 800;

// Sollen Fehler anzeigt (Wert: true) werden oder unterdrückt (Wert: false) (erlaubte Werte: "true" oder "false")
var displayError = true;


// AB HIER LÄSST DU BESSER DEINE PFOTEN VOM SKRIPT ...
// ... AUSSER DU HAST AHNUNG WAS DU DA MACHST

// Global special config
var cacheName = 'profileCache';
var profileOpenDelay = 75;

// Global vars
var activeProfile = null;
var user = undefined;

// Modify template before rendering
modifyTemplates();

// Do general stuff
$(document).ready(function() {
    // Build cache
    createProfileCache();
    
    // CSS code
    var css = '\
.user.profile { display: inline; } \
.user.profile > .name { cursor: pointer; } \
.user.profile > .name.active { text-decoration: underline; } \
.user.profile > .extended { z-index: 9000; display: none; font-size: 12px; position: absolute; margin-top: 5px; padding: 8px; background-color: rgba(0, 0, 0, 0.8); } \
.user.profile > .extended > ul { list-style: none; padding: 0px; margin: 0px; } \
.user.profile > .extended > ul > li { display: inline; white-space: nowrap; } \
.user.profile > .extended > ul > li > span { } \
.user.profile > .extended > ul > li > span:after { clear: both; display: block; content: " "; padding-bottom: 4px;} \
.user.profile > .extended > ul > li > span > span { } \
.user.profile > .extended > ul > li > span > span:first-child { float: left; margin-right: 10px; color: #BBBBBB; } \
.user.profile > .extended > ul > li > span > span:last-child { float: right; margin-left: 10px; color: #EE4D2E; } \
.user.profile > .extended > ul > li > span > a { margin-top: 7px; margin-left: 3px; margin-right: 3px; padding-left: 17px; padding-right: 17px; } \
.user.profile > .extended > ul > li > span > a[data-action="follow"][data-follow-state="1"] { background-color: #555555; } \
.user.profile > .extended > ul > li > span > a[data-action="follow"][data-follow-state="1"]:hover { background-color: #F5F7F6; } \
.user.profile > .extended > ul > li > span > a:first-child { margin-left: 0px; } \
.user.profile > .extended > ul > li > span > a:last-child { margin-right: 0px; } \
.user.profile > .extended > ul > li > ul { list-style: none; padding: 0px; margin: 0px; margin-top: 3px; text-align: right; } \
.user.profile > .extended > ul > li > ul > li { display: inline; } \
.comments > .comment-box:last-child { margin-bottom: 105px; } \
'; // #EE4D2E + #F5F7F6
    
    // Now bind action to item and comment
    $(document).on('click', 'div.user.profile[data-showable][data-visible] > span.name', function() { p.navigateTo('user/' + this.innerHTML, p.NAVIGATE.DEFAULT); });
    $(document).on('mouseover', 'div.user.profile[data-showable][data-visible]', function() { handleUserHoverProfile(this, 'request') });
    $(document).on('mouseout', 'div.user.profile[data-showable][data-visible]', function() { handleUserHoverProfile(this, 'close') });
    
    // Then bind follow action
    $(document).on('click', 'a.confirm-button[data-action="follow"]', function() { followUser(this); });
    
    // CSS stuff
    $('head').append('<style type="text/css">' + css + '</style>');
});

function hasPr0mium() {
    return p.user.paid;
}

function modifyTemplates() {
    // Modify stream template (item and comment)
    refreshTemplate(p.View.Stream.Item, 'item.user', 'item.mark');
    refreshTemplate(p.View.Stream.Comments, 'c.name', 'c.mark');
    
    // Modify stelz template
    refreshTemplate(p.View.FollowList, 'followed.name', 'followed.mark');
    
    // Modify inbox template (thread, sender and comment)
    refreshTemplate(p.View.Inbox, 't.partner.name', 't.partner.mark');
    refreshTemplate(p.View.Inbox, 'm.senderName', 'm.senderMark');
    refreshTemplate(p.View.Inbox, 'm.name', 'm.mark');
}

function refreshTemplate(view, name, mark) {
    // Build regex and profile
    var regex = buildProfileRegex(name, mark);
    var update= buildProfileTemplate(name, mark);
    
    // Replace in template
    view.prototype.template = view.prototype.template.replace(regex, update);
}

function buildProfileRegex(name, mark) {
    return (new RegExp('<a href="#user/{' + name + '}" class="user um{' + mark + '}">{' + name + '}</a>', 'g'));
}

function buildProfileTemplate(name, mark) {
    return '<div class="user profile um{' + mark + '}" data-showable="0" data-visible="0"><span class="name">{' + name + '}</span>\
<div class="extended" data-user="{' + name + '}">\
  <ul>\
    <li style="text-align: center">\
      <span>\
        <a href="#user/{' + name + '}" class="confirm-button"><span class="pict"></span>Profil</a>\
        <a href="#user/{' + name + '}/uploads" class="confirm-button"><span class="pict"></span>Uploads</a>\
        <a class="confirm-button" data-user="{' + name + '}" data-action="follow" data-follow-state="0"><span class="pict">@</span> <span class="description">stelz</span></a>\
      </span>\
    </li>\
  </ul>\
</div>\
</div>';
}

function createProfileCache() {
    $('body').append('<div id="' + cacheName + '" hidden="hidden"></div>');
}

function isProfileCached(user) {
    return ($('body > div#' + cacheName + ' > div[data-user="' + user + '"]').length > 0);
}

function changeFollowState(user, newState) {
    // Change all buttons
    $(document).find('a[data-action="follow"][data-user="' + user + '"]').each(function() {
        // Change attribute and text
        $(this).attr('data-follow-state', (newState ? '1' : '0'));
        $(this).children('span.description').text((newState ? 'unstelz' : 'stelz'));
    });
}

function followUser(button) {
    // Handle login
    if (!p.mainView.requireLogin()) {
        return false;
    }
    
    // Handle pr0mium
    if (!hasPr0mium()) {
        p.navigateTo('pr0mium', p.NAVIGATE.DEFAULT);
    }
    
    // Get real button and handle follow action
    button = $('body').find(button);
    var state = button.attr('data-follow-state') == '1';
    var action = (state ? 'un' : '') + 'follow';
    var user = button.attr('data-user');
    
    // Now talk to api ...
    p.api.post('profile.' + action, {
        name: user
    }, function(response) {
        // ... then display change
        changeFollowState(user, !state);
    }, (displayError ? function() { console.log('Fehler beim Stelzen'); } : p.ServerAPI.SILENT));
    
    // Return
    return true;
}

function addProfileInCache(user, data, badges) {
    // Prepare final output
    var content = '<div data-type="cache" data-user="' + user + '">';
    
    // Handle data
    for (var i = 0; i < data.length; i++) {
        content = content + createCacheLine(data[i].name, data[i].value);
    }
    
    // Handle badges
    if (badges.length) {
        var renderedBadges = '';
        for (var i = 0; i < badges.length; i++) {
            renderedBadges = renderedBadges + '<li><a class="badge" href="http://' + CONFIG.HOST + '/' + badges[i].link.substr(1) + '" title="' + badges[i].description + '"><img src="/media/badges/' + badges[i].image + '" class="badge" alt="" />' + (badges[i].extra != undefined ? '<span class="badge-extra' + (badges[i].css != undefined ? ' badge-' + badges[i].css : '') + '">' + badges[i].extra + '</span>' : '') + '</a></li>';
        }
        
        content = content + createCacheList(renderedBadges);
    }
    
    // Finish final line
    content = content + '</div>';
    
    // Provide in cache
    $('body > div#' + cacheName).append(content);
}

function createCacheList(content) {
    return '<li><ul>' + content + '</ul></li>';
}

function createCacheLine(left, right) {
    return '<li><span><span>' + left + '</span> <span>' + right + '</span></span></li>';
}

function loadProfileFromCache(user, target) {
    // Try to load from cache
    var cache = $('body > div#' + cacheName + ' > div[data-type="cache"][data-user="' + user + '"]');
    if (cache.length > 0) {
        target.children('div.extended').children('ul').children('li:first-child').before(cache.html());
        target.attr('data-showable', '1');
    }
}

function handleLoadedData(data, node) {
    // Handle data
    var profile = [];
    var user = data.user;
    var badges = data.badges;
    
    // Rank
    if (user.mark != 4) {
        profile.push({
            name: 'Status',
            value: p.User.MARK[user.mark]
        });
    }
    
    // Score and registration
    profile.push({
        name: 'Benis',
        value: user.score
    }, {
        name: (user.score > 0 ? 'gewachsen seit' : 'geschrumpft seit'),
        value: (new Date (user.registered * 1000)).relativeTime(true, true)
    });
    
    // Bann
    if (user.banned) {
        // Handle duration
        var duration = '';
        if (user.bannedUntil == null) {
            duration = 'immer';
        } else {
            duration = new Date(user.bannedUntil * 1000);
            duration = duration.relativeTime(false, true);
        }
        
        // Add to profile
        profile.push({
            name: 'gesperrt f&uuml;r',
            value: duration
        });
    }
    
    // Static comment badge
    for (var i = 0; i < p.View.User.BADGE.COMMENTS.COUNTS.length; i++) {
        if (data.commentCount > p.View.User.BADGE.COMMENTS.COUNTS[i].count) {
            var badge = p.copy(p.View.User.BADGE.COMMENTS.TEMPLATE);
            badge.description = badge.description.replace(/%c/g, p.View.User.BADGE.COMMENTS.COUNTS[i].count);
            badge.link = '#user/' + data.user.name + '/comments/before/' + (data.comments.length ? data.comments.first().created + 1 : 0);
            badge.extra = p.View.User.BADGE.COMMENTS.COUNTS[i].name;
            badge.css = 'comments';
            
            data.badges.push(badge);
        }
    }
    
    // Static year badge
    var years = Math.floor((Date.now() - data.user.registered * 1000) / (365 * 24 * 60 * 60 * 1000));
    if (years > 0) {
        var badge = p.copy(p.View.User.BADGE.YEARS.TEMPLATE);
        badge.description = badge.description.replace(/%y/g, 'Jahr'.inflect(years));
        badge.link = '#user/' + data.user.name;
        badge.extra = years.toString();
        badge.css = 'years';
        
        data.badges.push(badge);
    }
    
    // Handle stelz button
    if (data.following && hasPr0mium()) {
        changeFollowState(user.name, true);
    }
    
    // Remove loading animation
    node.children('div.extended').children('div.loader').remove();
    node.children('div.extended').children('ul').css('display', '');
    
    // Provide to cache and then use
    addProfileInCache(user.name, profile, data.badges);
    loadProfileFromCache(user.name, node);
}

function loadProfileFromAPI(user, node) {
    // Just do one api request
    if (node.attr('data-request-done') != undefined) {
        return;
    }
    
    // Mark as done
    node.attr('data-request-done', '1');
    
    // Add loading animation
    node.children('div.extended').children('ul').css('display', 'none');
    node.children('div.extended').append(p.View.Base.LoadingAnimHTML);
    node.children('div.extended').children('div.loader').css('margin-top', '32px');
    
    // Start ajax request
    p.api.get('profile.info', {
        name: user
    }, function(response) {
        handleLoadedData(response, node);
    }, (displayError ? function() { console.log('Fehler beim Profil laden'); } : p.ServerAPI.SILENT));
}

function handleUserHoverProfile(element, action, force) {
    // Mouse hovered profile link
    if (action == 'request') {
        // Get html node and control existance
        element = $('body').find(element);
        if (!element.length) {
            return;
        }
        
        // Set pointer
        activeProfile = element;
        
        // Show that we will open it
        element.children('span.name').addClass('active');
        
        // Start open timer
        setTimeout(function() { handleUserHoverProfile(element, 'open') }, profileOpenDelay);
        
    // Control if mouse still over profile
    } else if (action === 'open') {
        // Just open if mouse still over and node exists
        if (element != activeProfile || !element.length) {
            return;
        }
        
        // Get open profile and close it
        var openProfile = $('div.user.profile[data-showable][data-visible=1]');
        if (openProfile.children('span.name').text() != element.children('span.name').text()) {
            handleUserHoverProfile(openProfile, 'timeout', true);
        }
        
        // Show visually and physically
        element.children('div.extended').show(profileAnimationSpeed);
        element.attr('data-visible', '1');
        
        // Load item data
        if (element.attr('data-showable') != 1) {
            // Get username
            var user = element.children('div.extended').attr('data-user');
            
            // Load from api
            if (!isProfileCached(user)) {
                loadProfileFromAPI(user, element);
            
            // Load from local cache
            } else {
                loadProfileFromCache(user, element);
            }
        }
        
    // Mouse left profile link / data
    } else if (action === 'close') {
        // Delete link to active profile
        activeProfile = null;
        
        // Set mouse leave timeout
        setTimeout(function() { handleUserHoverProfile(element, 'timeout') }, profileMinShowDuration);
        
    // Profile will be closed cause no active view
    } else if (action === 'timeout') {
        // Get html node
        element = $('body').find(element);
        
        // Control node existance
        if (!element.length) {
            return;
        }
        
        // Just close if not still open
        if ((activeProfile == null && activeProfile != element) || force == true) {
            // Close physically ...
            element.attr('data-visible', '0');
            
            // ... then visually
            element.children('span.name').removeClass('active');
            element.children('div.extended').hide(profileAnimationSpeed);
        }
    }
}
